{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "79a74e6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from DNABERT.motif import motif_utils\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b8ad7973",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "seq_df shape (15, 1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>opJS2_1xTetO</th>\n",
       "      <td>ATGTGTGAGTTTTGTCGGATCCGAGACGACCGCCTGATCCGCCCGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>opJS2_2xTetO</th>\n",
       "      <td>ATGTGTGAGTTTTGTCGGATCCGAGACGACCGCCTGATCCGCCCGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>opJS2_3xTetO</th>\n",
       "      <td>ATGTGTGAGTTTTGTCGGATCCGAGACGACCGCCTGATCCGCCCGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>opJS2_4xTetO</th>\n",
       "      <td>ATGTGTGAGTTTTGTCGGATCCGAGACGACCGCCTGATCCGCCCGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>opJS2_5xTetO</th>\n",
       "      <td>ATGTGTGAGTTTTGTCGGATCCGAGACGACCGCCTGATCCGCCCGT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                            seq\n",
       "ID                                                             \n",
       "opJS2_1xTetO  ATGTGTGAGTTTTGTCGGATCCGAGACGACCGCCTGATCCGCCCGT...\n",
       "opJS2_2xTetO  ATGTGTGAGTTTTGTCGGATCCGAGACGACCGCCTGATCCGCCCGT...\n",
       "opJS2_3xTetO  ATGTGTGAGTTTTGTCGGATCCGAGACGACCGCCTGATCCGCCCGT...\n",
       "opJS2_4xTetO  ATGTGTGAGTTTTGTCGGATCCGAGACGACCGCCTGATCCGCCCGT...\n",
       "opJS2_5xTetO  ATGTGTGAGTTTTGTCGGATCCGAGACGACCGCCTGATCCGCCCGT..."
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_file = \"data/opJS.amplicon.long.fa\"\n",
    "\n",
    "reverse_complement = lambda x: str(Seq(x.seq).reverse_complement())\n",
    "\n",
    "\n",
    "with open(sequence_file) as fasta_file:\n",
    "    identifiers = []\n",
    "    seqs = []\n",
    "    for seq_record in SeqIO.parse(fasta_file, 'fasta'):\n",
    "        identifiers.append(seq_record.id)\n",
    "        seq_ = reverse_complement(seq_record) # ! we reverse complement for this experiment\n",
    "        seqs.append(str(seq_))\n",
    "\n",
    "#converting lists to pandas Series    \n",
    "s1 = pd.Series(identifiers, name='fragment')\n",
    "s2 = pd.Series(seqs, name='seq')\n",
    "#Gathering Series into a pandas DataFrame and rename index as ID column\n",
    "seq_df = pd.DataFrame(dict(ID=s1, seq=s2)).set_index(['ID'])\n",
    "\n",
    "print(f\"seq_df shape {seq_df.shape}\")\n",
    "seq_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8f7ce25c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['opJS2_1xTetO', 'opJS2_2xTetO', 'opJS2_3xTetO', 'opJS2_4xTetO',\n",
       "       'opJS2_5xTetO', 'opJS2_6xTetO', 'opJS2_7xTetO', 'opJS2_7xTetO-compact',\n",
       "       'opJS2_6xTetO-compact', 'opJS2_5xTetO-compact', 'opJS2_6xTetO-spread',\n",
       "       'opJS2_5xTetO-spread', 'opJS2_4xTetO-spread', 'opJS2_3xTetO-spread',\n",
       "       'opJS2_2xTetO-spread'],\n",
       "      dtype='object', name='ID')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_df.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f6154fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "KMER = 3\n",
    "\n",
    "def process_sequence(seq):\n",
    "    seq = str(Seq(seq).reverse_complement()) # reverse-complement the sequence\n",
    "    return motif_utils.seq2kmer(seq,KMER) # split sequence into k-mers for bert\n",
    "\n",
    "#conditions\n",
    "name_c = lambda group,dox,mtase,constr: f\"{group}_opJS2-{dox}{mtase}_LONG.opJS2_{constr}.full_unclustered.matrix\"\n",
    "group_c = [\"L057\",\"L058\",\"L059\",\"L060\"]\n",
    "dox_c = [\"24hr-1000ngDox\",\"no-dox\"]\n",
    "mtase_c = [\"\",\"-plusCpG\",]\n",
    "construct_c = [f\"{i}xTetO\" for i in range (1,8)]\n",
    "\n",
    "def process_construct(group,dox,mtase,construct):\n",
    "    experiment_file = name_c(group,dox,mtase,construct)\n",
    "    print(experiment_file)\n",
    "    file_path = f\"data/{experiment_file}\"\n",
    "    data = pd.read_csv(file_path,sep='\\t', header=[0])\n",
    "    data = data.iloc[: , 1:] # remove index column\n",
    "    return np.array(data.mode(0).iloc[0]) # get consensus status across reads, return it as array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2dd7ed83",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_condition(dox,mtase):\n",
    "    data = []\n",
    "    for construct in construct_c:\n",
    "        for group in group_c:\n",
    "            try:\n",
    "                mask = process_construct(group,dox,mtase,construct)\n",
    "                sequence = seq_df.loc[f\"opJS2_{construct}\"].seq.lower()\n",
    "                data.append([group,construct,sequence,mask])\n",
    "            except:\n",
    "                continue\n",
    "    return pd.DataFrame(data, columns=['group','construct', 'sequence','label'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f30e947e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "L057_opJS2-24hr-1000ngDox_LONG.opJS2_1xTetO.full_unclustered.matrix\n",
      "L058_opJS2-24hr-1000ngDox_LONG.opJS2_1xTetO.full_unclustered.matrix\n",
      "L059_opJS2-24hr-1000ngDox_LONG.opJS2_1xTetO.full_unclustered.matrix\n",
      "L060_opJS2-24hr-1000ngDox_LONG.opJS2_1xTetO.full_unclustered.matrix\n",
      "L057_opJS2-24hr-1000ngDox_LONG.opJS2_2xTetO.full_unclustered.matrix\n",
      "L058_opJS2-24hr-1000ngDox_LONG.opJS2_2xTetO.full_unclustered.matrix\n",
      "L059_opJS2-24hr-1000ngDox_LONG.opJS2_2xTetO.full_unclustered.matrix\n",
      "L060_opJS2-24hr-1000ngDox_LONG.opJS2_2xTetO.full_unclustered.matrix\n",
      "L057_opJS2-24hr-1000ngDox_LONG.opJS2_3xTetO.full_unclustered.matrix\n",
      "L058_opJS2-24hr-1000ngDox_LONG.opJS2_3xTetO.full_unclustered.matrix\n",
      "L059_opJS2-24hr-1000ngDox_LONG.opJS2_3xTetO.full_unclustered.matrix\n",
      "L060_opJS2-24hr-1000ngDox_LONG.opJS2_3xTetO.full_unclustered.matrix\n",
      "L057_opJS2-24hr-1000ngDox_LONG.opJS2_4xTetO.full_unclustered.matrix\n",
      "L058_opJS2-24hr-1000ngDox_LONG.opJS2_4xTetO.full_unclustered.matrix\n",
      "L059_opJS2-24hr-1000ngDox_LONG.opJS2_4xTetO.full_unclustered.matrix\n",
      "L060_opJS2-24hr-1000ngDox_LONG.opJS2_4xTetO.full_unclustered.matrix\n",
      "L057_opJS2-24hr-1000ngDox_LONG.opJS2_5xTetO.full_unclustered.matrix\n",
      "L058_opJS2-24hr-1000ngDox_LONG.opJS2_5xTetO.full_unclustered.matrix\n",
      "L059_opJS2-24hr-1000ngDox_LONG.opJS2_5xTetO.full_unclustered.matrix\n",
      "L060_opJS2-24hr-1000ngDox_LONG.opJS2_5xTetO.full_unclustered.matrix\n",
      "L057_opJS2-24hr-1000ngDox_LONG.opJS2_6xTetO.full_unclustered.matrix\n",
      "L058_opJS2-24hr-1000ngDox_LONG.opJS2_6xTetO.full_unclustered.matrix\n",
      "L059_opJS2-24hr-1000ngDox_LONG.opJS2_6xTetO.full_unclustered.matrix\n",
      "L060_opJS2-24hr-1000ngDox_LONG.opJS2_6xTetO.full_unclustered.matrix\n",
      "L057_opJS2-24hr-1000ngDox_LONG.opJS2_7xTetO.full_unclustered.matrix\n",
      "L058_opJS2-24hr-1000ngDox_LONG.opJS2_7xTetO.full_unclustered.matrix\n",
      "L059_opJS2-24hr-1000ngDox_LONG.opJS2_7xTetO.full_unclustered.matrix\n",
      "L060_opJS2-24hr-1000ngDox_LONG.opJS2_7xTetO.full_unclustered.matrix\n",
      "L057_opJS2-no-dox_LONG.opJS2_1xTetO.full_unclustered.matrix\n",
      "L058_opJS2-no-dox_LONG.opJS2_1xTetO.full_unclustered.matrix\n",
      "L059_opJS2-no-dox_LONG.opJS2_1xTetO.full_unclustered.matrix\n",
      "L060_opJS2-no-dox_LONG.opJS2_1xTetO.full_unclustered.matrix\n",
      "L057_opJS2-no-dox_LONG.opJS2_2xTetO.full_unclustered.matrix\n",
      "L058_opJS2-no-dox_LONG.opJS2_2xTetO.full_unclustered.matrix\n",
      "L059_opJS2-no-dox_LONG.opJS2_2xTetO.full_unclustered.matrix\n",
      "L060_opJS2-no-dox_LONG.opJS2_2xTetO.full_unclustered.matrix\n",
      "L057_opJS2-no-dox_LONG.opJS2_3xTetO.full_unclustered.matrix\n",
      "L058_opJS2-no-dox_LONG.opJS2_3xTetO.full_unclustered.matrix\n",
      "L059_opJS2-no-dox_LONG.opJS2_3xTetO.full_unclustered.matrix\n",
      "L060_opJS2-no-dox_LONG.opJS2_3xTetO.full_unclustered.matrix\n",
      "L057_opJS2-no-dox_LONG.opJS2_4xTetO.full_unclustered.matrix\n",
      "L058_opJS2-no-dox_LONG.opJS2_4xTetO.full_unclustered.matrix\n",
      "L059_opJS2-no-dox_LONG.opJS2_4xTetO.full_unclustered.matrix\n",
      "L060_opJS2-no-dox_LONG.opJS2_4xTetO.full_unclustered.matrix\n",
      "L057_opJS2-no-dox_LONG.opJS2_5xTetO.full_unclustered.matrix\n",
      "L058_opJS2-no-dox_LONG.opJS2_5xTetO.full_unclustered.matrix\n",
      "L059_opJS2-no-dox_LONG.opJS2_5xTetO.full_unclustered.matrix\n",
      "L060_opJS2-no-dox_LONG.opJS2_5xTetO.full_unclustered.matrix\n",
      "L057_opJS2-no-dox_LONG.opJS2_6xTetO.full_unclustered.matrix\n",
      "L058_opJS2-no-dox_LONG.opJS2_6xTetO.full_unclustered.matrix\n",
      "L059_opJS2-no-dox_LONG.opJS2_6xTetO.full_unclustered.matrix\n",
      "L060_opJS2-no-dox_LONG.opJS2_6xTetO.full_unclustered.matrix\n",
      "L057_opJS2-no-dox_LONG.opJS2_7xTetO.full_unclustered.matrix\n",
      "L058_opJS2-no-dox_LONG.opJS2_7xTetO.full_unclustered.matrix\n",
      "L059_opJS2-no-dox_LONG.opJS2_7xTetO.full_unclustered.matrix\n",
      "L060_opJS2-no-dox_LONG.opJS2_7xTetO.full_unclustered.matrix\n",
      "L057_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_1xTetO.full_unclustered.matrix\n",
      "L058_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_1xTetO.full_unclustered.matrix\n",
      "L059_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_1xTetO.full_unclustered.matrix\n",
      "L060_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_1xTetO.full_unclustered.matrix\n",
      "L057_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_2xTetO.full_unclustered.matrix\n",
      "L058_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_2xTetO.full_unclustered.matrix\n",
      "L059_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_2xTetO.full_unclustered.matrix\n",
      "L060_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_2xTetO.full_unclustered.matrix\n",
      "L057_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_3xTetO.full_unclustered.matrix\n",
      "L058_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_3xTetO.full_unclustered.matrix\n",
      "L059_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_3xTetO.full_unclustered.matrix\n",
      "L060_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_3xTetO.full_unclustered.matrix\n",
      "L057_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_4xTetO.full_unclustered.matrix\n",
      "L058_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_4xTetO.full_unclustered.matrix\n",
      "L059_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_4xTetO.full_unclustered.matrix\n",
      "L060_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_4xTetO.full_unclustered.matrix\n",
      "L057_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_5xTetO.full_unclustered.matrix\n",
      "L058_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_5xTetO.full_unclustered.matrix\n",
      "L059_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_5xTetO.full_unclustered.matrix\n",
      "L060_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_5xTetO.full_unclustered.matrix\n",
      "L057_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_6xTetO.full_unclustered.matrix\n",
      "L058_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_6xTetO.full_unclustered.matrix\n",
      "L059_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_6xTetO.full_unclustered.matrix\n",
      "L060_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_6xTetO.full_unclustered.matrix\n",
      "L057_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_7xTetO.full_unclustered.matrix\n",
      "L058_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_7xTetO.full_unclustered.matrix\n",
      "L059_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_7xTetO.full_unclustered.matrix\n",
      "L060_opJS2-24hr-1000ngDox-plusCpG_LONG.opJS2_7xTetO.full_unclustered.matrix\n",
      "L057_opJS2-no-dox-plusCpG_LONG.opJS2_1xTetO.full_unclustered.matrix\n",
      "L058_opJS2-no-dox-plusCpG_LONG.opJS2_1xTetO.full_unclustered.matrix\n",
      "L059_opJS2-no-dox-plusCpG_LONG.opJS2_1xTetO.full_unclustered.matrix\n",
      "L060_opJS2-no-dox-plusCpG_LONG.opJS2_1xTetO.full_unclustered.matrix\n",
      "L057_opJS2-no-dox-plusCpG_LONG.opJS2_2xTetO.full_unclustered.matrix\n",
      "L058_opJS2-no-dox-plusCpG_LONG.opJS2_2xTetO.full_unclustered.matrix\n",
      "L059_opJS2-no-dox-plusCpG_LONG.opJS2_2xTetO.full_unclustered.matrix\n",
      "L060_opJS2-no-dox-plusCpG_LONG.opJS2_2xTetO.full_unclustered.matrix\n",
      "L057_opJS2-no-dox-plusCpG_LONG.opJS2_3xTetO.full_unclustered.matrix\n",
      "L058_opJS2-no-dox-plusCpG_LONG.opJS2_3xTetO.full_unclustered.matrix\n",
      "L059_opJS2-no-dox-plusCpG_LONG.opJS2_3xTetO.full_unclustered.matrix\n",
      "L060_opJS2-no-dox-plusCpG_LONG.opJS2_3xTetO.full_unclustered.matrix\n",
      "L057_opJS2-no-dox-plusCpG_LONG.opJS2_4xTetO.full_unclustered.matrix\n",
      "L058_opJS2-no-dox-plusCpG_LONG.opJS2_4xTetO.full_unclustered.matrix\n",
      "L059_opJS2-no-dox-plusCpG_LONG.opJS2_4xTetO.full_unclustered.matrix\n",
      "L060_opJS2-no-dox-plusCpG_LONG.opJS2_4xTetO.full_unclustered.matrix\n",
      "L057_opJS2-no-dox-plusCpG_LONG.opJS2_5xTetO.full_unclustered.matrix\n",
      "L058_opJS2-no-dox-plusCpG_LONG.opJS2_5xTetO.full_unclustered.matrix\n",
      "L059_opJS2-no-dox-plusCpG_LONG.opJS2_5xTetO.full_unclustered.matrix\n",
      "L060_opJS2-no-dox-plusCpG_LONG.opJS2_5xTetO.full_unclustered.matrix\n",
      "L057_opJS2-no-dox-plusCpG_LONG.opJS2_6xTetO.full_unclustered.matrix\n",
      "L058_opJS2-no-dox-plusCpG_LONG.opJS2_6xTetO.full_unclustered.matrix\n",
      "L059_opJS2-no-dox-plusCpG_LONG.opJS2_6xTetO.full_unclustered.matrix\n",
      "L060_opJS2-no-dox-plusCpG_LONG.opJS2_6xTetO.full_unclustered.matrix\n",
      "L057_opJS2-no-dox-plusCpG_LONG.opJS2_7xTetO.full_unclustered.matrix\n",
      "L058_opJS2-no-dox-plusCpG_LONG.opJS2_7xTetO.full_unclustered.matrix\n",
      "L059_opJS2-no-dox-plusCpG_LONG.opJS2_7xTetO.full_unclustered.matrix\n",
      "L060_opJS2-no-dox-plusCpG_LONG.opJS2_7xTetO.full_unclustered.matrix\n"
     ]
    }
   ],
   "source": [
    "# different conditions should have different dataframes\n",
    "df_naming = {\"24hr-1000ngDox\":'dox','no-dox':'nodox',\n",
    "            \"\":\"nocpg\",\"-plusCpG\":\"cpg\"}\n",
    "for mtase in mtase_c:\n",
    "    for dox in dox_c:\n",
    "        name = f'{df_naming[mtase]}_{df_naming[dox]}'\n",
    "        globals()[name] = process_condition(dox,mtase)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "98180554",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>construct</th>\n",
       "      <th>sequence</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>L059</td>\n",
       "      <td>1xTetO</td>\n",
       "      <td>atgtgtgagttttgtcggatccgagacgaccgcctgatccgcccgt...</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L059</td>\n",
       "      <td>2xTetO</td>\n",
       "      <td>atgtgtgagttttgtcggatccgagacgaccgcctgatccgcccgt...</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L059</td>\n",
       "      <td>3xTetO</td>\n",
       "      <td>atgtgtgagttttgtcggatccgagacgaccgcctgatccgcccgt...</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>L059</td>\n",
       "      <td>4xTetO</td>\n",
       "      <td>atgtgtgagttttgtcggatccgagacgaccgcctgatccgcccgt...</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>L059</td>\n",
       "      <td>5xTetO</td>\n",
       "      <td>atgtgtgagttttgtcggatccgagacgaccgcctgatccgcccgt...</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>L059</td>\n",
       "      <td>6xTetO</td>\n",
       "      <td>atgtgtgagttttgtcggatccgagacgaccgcctgatccgcccgt...</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>L059</td>\n",
       "      <td>7xTetO</td>\n",
       "      <td>atgtgtgagttttgtcggatccgagacgaccgcctgatccgcccgt...</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  group construct                                           sequence  \\\n",
       "0  L059    1xTetO  atgtgtgagttttgtcggatccgagacgaccgcctgatccgcccgt...   \n",
       "1  L059    2xTetO  atgtgtgagttttgtcggatccgagacgaccgcctgatccgcccgt...   \n",
       "2  L059    3xTetO  atgtgtgagttttgtcggatccgagacgaccgcctgatccgcccgt...   \n",
       "3  L059    4xTetO  atgtgtgagttttgtcggatccgagacgaccgcctgatccgcccgt...   \n",
       "4  L059    5xTetO  atgtgtgagttttgtcggatccgagacgaccgcctgatccgcccgt...   \n",
       "5  L059    6xTetO  atgtgtgagttttgtcggatccgagacgaccgcctgatccgcccgt...   \n",
       "6  L059    7xTetO  atgtgtgagttttgtcggatccgagacgaccgcctgatccgcccgt...   \n",
       "\n",
       "                                               label  \n",
       "0  [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -...  \n",
       "1  [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -...  \n",
       "2  [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -...  \n",
       "3  [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -...  \n",
       "4  [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -...  \n",
       "5  [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -...  \n",
       "6  [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nocpg_dox"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecefefae",
   "metadata": {},
   "source": [
    "## Process for BERT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "05a87c05",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "\n",
    "def arr2kmer(seq, k):\n",
    "    \"\"\"\n",
    "    Convert original sequence to kmers\n",
    "    \n",
    "    Arguments:\n",
    "    seq -- str, original sequence.\n",
    "    k -- int, kmer of length k specified.\n",
    "    \n",
    "    Returns:\n",
    "    kmers -- str, kmers separated by space\n",
    "    \"\"\"\n",
    "    kmer = [list(seq[x:x+k]) for x in range(len(seq)+1-k)]\n",
    "    return kmer\n",
    "\n",
    "context_dict = {}\n",
    "c0 = 2\n",
    "for c1 in ['cpg','nocpg']:\n",
    "    for c2 in ['dox','nodox']:\n",
    "        context_dict[f'{c1}_{c2}'] = c0\n",
    "        c0 +=1\n",
    "        \n",
    "def add_context_vector(label,condition,how='np'):\n",
    "    # 2 ways to add context vector - new position (np) and concat\n",
    "    context = context_dict[condition]\n",
    "    if how == 'np':\n",
    "        label_ = []\n",
    "        for kmer in label:\n",
    "            kmer.append(context)\n",
    "            label_.append(kmer)\n",
    "        return label_\n",
    "    else: # concat\n",
    "        context = [context_dict[condition] for _ in range(KMER)]\n",
    "        label_ = []\n",
    "        for kmer in label:\n",
    "            label_.append([kmer,context])\n",
    "        return label_\n",
    "        \n",
    "def prep_for_dnabert(df,condition):\n",
    "    for i,row in df.iterrows():\n",
    "        df.at[i,'sequence'] = process_sequence(row.sequence)\n",
    "        #df.at[i,'label'] = str(random.choice([0, 1])) #wrong,just trying binary labels\n",
    "        df.at[i,'label'] = arr2kmer(row.label, KMER)\n",
    "        df.at[i,'label'] = add_context_vector(row.label,condition)\n",
    "    return df[['sequence','label']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "05af838a",
   "metadata": {},
   "outputs": [],
   "source": [
    "nocpg_dox_bert = prep_for_dnabert(nocpg_dox,'nocpg_dox')\n",
    "nocpg_nodox_bert = prep_for_dnabert(nocpg_nodox,'nocpg_nodox')\n",
    "cpg_dox_bert = prep_for_dnabert(cpg_dox,'cpg_dox')\n",
    "cpg_nodox_bert = prep_for_dnabert(cpg_nodox,'cpg_nodox')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4a907e09",
   "metadata": {},
   "outputs": [],
   "source": [
    "nocpg_nodox_bert.to_csv(f'DNABERT/sample_data/{KMER}/train.tsv',index=False,sep=\"\\t\")\n",
    "nocpg_nodox_bert.to_csv(f'DNABERT/sample_data/{KMER}/dev.tsv',index=False,sep=\"\\t\") # need for evaluation, TODO - in future switch to new data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34031822",
   "metadata": {},
   "source": [
    "## Save files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4ba1446b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "folder_processed = 'processed_data'\n",
    "#os.mkdir(folder_processed)\n",
    "#os.mkdir(f'{folder_processed}/{KMER}')\n",
    "\n",
    "for c1 in ['cpg','nocpg']:\n",
    "    for c2 in ['dox','nodox']:\n",
    "        name1 = f'{c1}_{c2}'\n",
    "        globals()[name1].to_csv(f'{folder_processed}/{name1}.csv',index=False)\n",
    "        name2 = f'{name1}_bert'\n",
    "        globals()[name2].to_csv(f'{folder_processed}/{KMER}/{name2}.tsv',sep=\"\\t\",)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd1b21e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cd examples\n",
    "\n",
    "# export KMER=6\n",
    "# export MODEL_PATH=6-new-12w-0\n",
    "# export DATA_PATH=sample_data/ft/$KMER\n",
    "# export OUTPUT_PATH=./ft/$KMER\n",
    "\n",
    "# python run_finetune.py \\\n",
    "#     --model_type dna \\\n",
    "#     --tokenizer_name=dna$KMER \\\n",
    "#     --model_name_or_path $MODEL_PATH \\\n",
    "#     --task_name dnaprom \\\n",
    "#     --do_train \\\n",
    "#     --do_eval \\\n",
    "#     --data_dir $DATA_PATH \\\n",
    "#     --max_seq_length 100 \\\n",
    "#     --per_gpu_eval_batch_size=32   \\\n",
    "#     --per_gpu_train_batch_size=32   \\\n",
    "#     --learning_rate 2e-4 \\\n",
    "#     --num_train_epochs 5.0 \\\n",
    "#     --output_dir $OUTPUT_PATH \\\n",
    "#     --evaluate_during_training \\\n",
    "#     --logging_steps 100 \\\n",
    "#     --save_steps 4000 \\\n",
    "#     --warmup_percent 0.1 \\\n",
    "#     --hidden_dropout_prob 0.1 \\\n",
    "#     --overwrite_output \\\n",
    "#     --weight_decay 0.01 \\\n",
    "#     --n_process 8"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dnabert",
   "language": "python",
   "name": "dnabert"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
